docs/design/reflection-flow.md

```diff
@@ -0,0 +1,189 @@
+# Reflection and IL Linker in .NET 5
+
+Unconstrained reflection in .NET presents a challenge for discovering parts of .NET apps that are (or are not) used. We had multiple attempts to solve the problem – probably the most complex one was in .NET Native. .NET Native was trying to make arbitrary reflection "just work" by using a combination of:
+
+* cross-method dataflow analysis framework (that could e.g. model the `dynamic` keyword in C#, or reason about cascading reflection patterns like `Type.GetType("Foo").GetMethod("Bar").ReturnType.GetMethod("Baz")`)
+* expressive annotation format (RD.XML) that could express things like "keep all properties and constructors on types passed as a parameter to this method, recursively". The idea was that library developers would annotate their libraries and things would "just work" for the user.
+
+Even with the level of investment in .NET Native it wasn't possible to make arbitrary reflection "just work" – before shipping, we had to make a decision to not do any tree-shaking on user assemblies by default because the reflection patterns were often (~20% of the Store app catalog) arbitrary enough that it wasn't possible to describe them in generic ways or detect them. The .NET Native compiler did not warn the user about presence of unrecognized patterns either because we expected there would be too much noise due to the disconnect between RD.XML (that simply described what to keep) and the dataflow analysis (that focused on reflection API usage).
+
+## Linker-safe reflection
+
+Note: While this document mostly talks about reflection, this includes reflection-like APIs with impact on linker's analysis such as `RuntimeHelpers.GetUninitializedObject`, `Marshal.PtrToStructure`, or `RuntimeHelpers.RunClassConstructor`.
+
+In .NET 5, we would like to carve out a _subset_ of reflection patterns that can be made safe in the presence of IL linker. Since IL linking is optional, users do not have to adhere to this subset. They can still use IL Linker if they do not adhere to this subset, but we will not provide guarantees that linking won't change the semantics of their app. Linker will warn if a reflection pattern within the app is not safe.
+
+To achieve safety, we'll logically classify methods into following categories:
+
+* Linker-safe: most code will fall into this category. Linking can be done safely based on information in the static callgraph.
+* Potentially unsafe: call to the method is unsafe if the linker cannot reason about a parameter value (e.g. `Type.GetType` with a type name string that could be unknown)
+* Always unsafe: calls to these methods are never safe (e.g. `Assembly.ExportedTypes`).
```

Comment:
I find this example to be an interesting one. 

Consider
```
int totalExportedTypes = 0;
foreach (var type in typeof(int).Assembly.ExportedTypes)
            {
                 totalExportedTypes++;
            }
if (totalExportedTypes != <some expected number>)
 throw;
```

`typeof(int)` is easily seen by the linker.  `Assembly` is easy.  and so is `ExportedTypes`.  The linker could mark all of the exported types.  The behavior of the linked assemblies would match the original program. 

So the linker could correctly mark all of the dependencies of the reflection API, however, doing so would result in over preserving in most real world use cases since I imagine inside that loop there would normally be some sort of filtering happening on the types.

Reply:
I think it's confusing to call anything just linker-unsafe - linker could keep the entire app and everything would be safe. So maybe a better way is to say that we choose not to support the pattern in the linker as we're not able to achieve good trimming results with it.

Code patterns like this example are typical in Dependency Injection (or similar) scenarios - I think that we can potentially provide different ways of solving that problem. But that's outside of this proposal.

Reply:
> confusing to call anything just linker-unsafe

Would it be better to talk about this linker-compatible/incompatible or linker-friendly/unfriendly?

Reply:
safe/unsafe meaning typically relates to security. This has not much to do with security.

Reply:
We've been using linker-friendly in some places already - so I'd be happy with that.

Reply:
> The linker could mark all of the exported types

It could, but it negates a lot of benefits of the linking (one unfortunate API call and suddenly none of the public types get linked away). It doesn't seem to be worth the effort to make this work.

The problem is also that once we have the type back, people are going to do other things with it (not just count them). So we either need to understand `IEnumerable<Type>` or just keep all members to be sure.

Reply:
"safe/unsafe meaning typically relates to security" I agree that it's good to not pollute the use of this convention, even though that's all it is.

Reply:
Yes, I agree, we wouldn't want to support the case of assembly getting types.

However, what about `typeof(Foo).GetFields()`?  in our linker we will mark all of the fields of `Foo` if we see this.  Our rational was, if you are getting the fields of the specific type you probably care about those fields and the odds of an undesired ripple effect are much lower.  I will say  I was/am on the fence about it.   I'm curious what stance you guys take.

Reply:
> However, what about `typeof(Foo).GetFields()`

I think keeping all fields on `Foo` in this case would align with how member access is treated in the proposed `MemberKinds` enum - type-level granularity feels more acceptable than assembly level granularity.

Reply:
I think the current proposal is to mark this with `AllFields` which would get us basically the same behavior as you already have. I personally think that we should be very careful about including more types than "necessary", but I'm much more comfortable including more members.